[WIP] New package: ly-0.5.2

cinerea0 opened the pull request:
[ly](https://github.com/nullgemm/ly) is a simple TUI display manager. It recently had a new tagged release, and I thought this would be a good time to try to package it. This PR would resolve #13583 and would allow #13649 to be closed. The template is based almost entirely on [this template](https://github.com/ericonr/void-packages/commit/eadabab675acaff062d138a2607d5920b125bfde) by @ericonr, and the service files were taken from [a fork of ly made to work with Void](https://github.com/drozdowsky/ly-void/commit/b972fe48a143e709dc695ac8e2c83dc57cee279f). The patches make minor adjustments to make ly display on the standard TTY 7 and remove the systemD service.

ly is able to build properly on my machine; I'll soon see if it build properly through travis. The problem with this package is that it doesn't seem to work quite right after it starts running. I get PAM errors when attempting to log in using ly; you can see those errors in [this issue in the ly repository](https://github.com/nullgemm/ly/issues/216). My hunch is that the problem lies in [ly's PAM file](https://github.com/nullgemm/ly/blob/master/res/pam.d/ly). Does anyone know how to patch this to work properly?

cinerea0 commented:
Could anyone who's familiar with C explain the error occuring on the ARM builds:
```
src/login.c:13:10: fatal error: security/pam_appl.h: No such file or directory

   13 | #include <security/pam_appl.h>

      |          ^~~~~~~~~~~~~~~~~~~~~

compilation terminated.
```
I think it means that there's a library that's not installed, but this error doesn't exist on any of the other architectures.

ericonr commented:
If you `vsed` away the `@`s in the Makefile, you can see what commands it's trying to run and try to debug from there. That looks like a cross compilation issue, though.

ericonr commented on srcpkgs/ly/patches/makefile.patch:
```diff
@@ -0,0 +1,27 @@
+Remove systemd service from install and uninstall targets
+--- makefile	2020-07-28 11:35:08.000000000 -0400
++++ "makefile copy"	2020-08-02 21:40:10.283778154 -0400
+@@ -78,7 +78,6 @@
+ 	@install -DZ $(RESD)/wsetup.sh -t $(DATADIR)
+ 	@install -dZ $(DATADIR)/lang
+ 	@install -DZ $(RESD)/lang/* -t $(DATADIR)/lang
+-	@install -DZ $(RESD)/ly.service -m 644 -t ${DESTDIR}/usr/lib/systemd/system
+ 	@install -DZ $(RESD)/pam.d/ly -m 644 -t ${DESTDIR}/etc/pam.d
+ 
+ installnoconf: $(BIND)/$(NAME)
+@@ -89,7 +88,6 @@
+ 	@install -DZ $(RESD)/wsetup.sh -t $(DATADIR)
+ 	@install -dZ $(DATADIR)/lang
+ 	@install -DZ $(RESD)/lang/* -t $(DATADIR)/lang
+-	@install -DZ $(RESD)/ly.service -m 644 -t ${DESTDIR}/usr/lib/systemd/system
+ 	@install -DZ $(RESD)/pam.d/ly -m 644 -t ${DESTDIR}/etc/pam.d
+ 
+ uninstall:
+@@ -97,7 +95,6 @@
+ 	@rm -rf ${DESTDIR}/etc/ly
+ 	@rm -rf $(DATADIR)
+ 	@rm -f ${DESTDIR}/usr/bin/ly
+-	@rm -f ${DESTDIR}/usr/lib/systemd/system/ly.service
+ 	@rm -f ${DESTDIR}/etc/pam.d/ly
```
Easier to just remove these in `post_install` instead of patching the makefile.

ericonr commented on srcpkgs/ly/patches/config.ini.patch:
```diff
@@ -0,0 +1,12 @@
+Change default tty to 7
+--- res/config.ini	2020-07-28 11:35:08.000000000 -0400
++++ "res/config copy.ini"	2020-08-02 21:37:52.572343702 -0400
+@@ -79,7 +79,7 @@
+ #term_reset_cmd = /usr/bin/tput reset
+ 
+ # tty in use
+-#tty = 2
++#tty = 7
+ 
+ # wayland setup command
+ #wayland_cmd = /etc/ly/wsetup.sh
```
Why default to 7?

cinerea0 replied:
7 is the tty that all (as far as I know) desktop environments launch on. With the previous default tty of 2, users would run into a scenario where they would log in via `ly` and receive no feedback that it had succeeded; they would need to manually switch to tty7 after logging in.

ericonr replied:
I see. Could we simply uncomment this from the config file, to avoid patching the source as well? We should open an issue/PR with upstream about making it a build time option, as well.

cinerea0 replied:
Yes, I can change the config file to avoid patching the source. Would the issue simply contain a request to add setting tty7 as the default as a build time option?

ericonr replied:
I would make it generic, so any tty can be chosen as default.

cinerea0 replied:
Issue opened: https://github.com/nullgemm/ly/issues/225

ericonr commented on srcpkgs/ly/files/ly/run:
```diff
@@ -0,0 +1,15 @@
+#!/bin/sh
+
+tty=${PWD##*-}
+
+[ -r conf ] && . ./conf
+
+if [ -x /sbin/getty -o -x /bin/getty ]; then
+	# busybox
+	GETTY=getty
+elif [ -x /sbin/agetty -o -x /bin/agetty ]; then
+	# util-linux
+	GETTY=agetty
+fi
+
+exec setsid ${GETTY} ${GETTY_ARGS} -nl /usr/bin/ly tty7 "${BAUD_RATE}" "${TERM_NAME}"
```
tty7 shouldn't be hardcoded.

cinerea0 replied:
What would be the correct way to set it? Subsitute `tty7` with `${tty}`?

ericonr replied:
no, because the service is being installed as `ly`. It would need to be installed as `ly-tty7` for the `tty=${PWD##*-}` trick to work.

cinerea0 replied:
It's hardcoded to tty2 in the original run file. I'm starting to think that I'm going to need to scrap a lot of the work done in the void fork on the service files.

cinerea0 replied:
Actually, now that I think about it, is it possible to read in data from another file in the conf file?

ericonr replied:
Ideally no, the conf file should be reasonably simple

ericonr replied:
What are you thinking of doing?

cinerea0 replied:
I'm considering `grep`ing the `/etc/ly/config.ini` file for the tty number and using that in the arguments passed to `/usr/bin/ly`.

cinerea0 replied:
As a side note, can I move the getty vs agetty thing in the run file into the conf file?

ericonr replied:
Honestly, I don't have any issue with making this package depend on `util-linux` directly, so imo you can just use `agetty`.

ericonr replied:
> I'm considering greping the /etc/ly/config.ini file for the tty number and using that in the arguments passed to /usr/bin/ly.

I prefer the user configuring this manually, with you leaving sensible defaults. Otherwise it becomes a little too magical.

cinerea0 replied:
Just changed it. Could you take a look at the conf, run, and config.ini.patch files to see if that's sufficient to allow users to understand what they need to do?

ericonr commented on srcpkgs/ly/files/ly/conf:
```diff
@@ -0,0 +1,14 @@
+	fi
+fi
+
+BAUD_RATE=38400
+TERM_NAME=linux
+
+StandardInput=tty
+TTYPath=/dev/tty7
+TTYReset=yes
+TTYVHangup=yes 
```
Are these being used anywhere?

cinerea0 replied:
BAUD_RATE and TERM_NAME are used in the `run` file. The others are inherited from the systemD service file. They don't appear to be used anywhere, but they do correspond to commands that should probably be added to the `run` file. In this case, the commands would be changing the virtual terminal to whichever one is specified there and issuing `reset` to clear the screen. At least, that's what I got from [this article](https://alan-mushi.github.io/2014/10/26/execute-an-interactive-script-at-boot-with-systemd.html).

ericonr replied:
Ok, so perhaps add a `reset` command before launching ly. Not sure what `hangup` gives us.

ericonr commented on srcpkgs/ly/INSTALL.msg:
```diff
@@ -0,0 +1 @@
+ly requires either xinit or sx to be available.
```
It doesn't, it can launch a shell or wayland session.

ericonr commented on srcpkgs/ly/template:
```diff
@@ -0,0 +1,64 @@
+build_style=gnu-makefile
+make_build_args="FLAGS=\"$CFLAGS\""
+conf_files="/etc/ly/config.ini"
+hostmakedepends="git"
```
Does it still need `git`?

cinerea0 replied:
No, but adding it shuts up the "git not found" output that pops up without it. Granted, you still get the "fatal, not a git repository" error with it installed, so I guess it doesn't matter. I'll remove it.

ericonr commented:
Try to add this patch as a commit https://github.com/nullgemm/ly/commit/8493c5a8ada16d3da56b23ff28791068c542d1eb for the PAM issue.

ericonr commented on srcpkgs/ly/template:
```diff
@@ -0,0 +1,64 @@
+}
+
+pre_build() {
+	exit
```
Why is this here?

cinerea0 replied:
No idea, I inherited it from a previous version of the template. I'll remove it and see what happens.

cinerea0 commented:
After using `vsed` to get rid of the `@`s at the start of lines, removing `git`, and removing that `exit`, I'm running into a compilation error I wasn't before:
```
building object obj/src/config.o
mkdir -p obj/src
cc -Isrc -Isub/ctypes -Isub/argoat/src -Isub/configator/src -Isub/dragonfail/src -Isub/termbox_next/src -fstack-clash-protection -D_FORTIFY_SOURCE=2 -mtune=generic -O2 -pipe    -c -o obj/src/config.o src/config.c
src/config.c: In function 'lang_load':
src/config.c:12:19: error: 'DATADIR' undeclared (first use in this function)
   12 |  #define INI_LANG DATADIR "/lang/%s.ini"
      |                   ^~~~~~~
src/config.c:141:22: note: in expansion of macro 'INI_LANG'
  141 |  snprintf(file, 256, INI_LANG, config.lang);
      |                      ^~~~~~~~
src/config.c:12:19: note: each undeclared identifier is reported only once for each function it appears in
   12 |  #define INI_LANG DATADIR "/lang/%s.ini"
      |                   ^~~~~~~
src/config.c:141:22: note: in expansion of macro 'INI_LANG'
  141 |  snprintf(file, 256, INI_LANG, config.lang);
      |                      ^~~~~~~~
src/config.c:12:27: error: expected ')' before string constant
   12 |  #define INI_LANG DATADIR "/lang/%s.ini"
      |                           ^~~~~~~~~~~~~~
src/config.c:141:22: note: in expansion of macro 'INI_LANG'
  141 |  snprintf(file, 256, INI_LANG, config.lang);
      |                      ^~~~~~~~
src/config.c: In function 'config_defaults':
src/config.c:292:30: error: 'DATADIR' undeclared (first use in this function)
  292 |  config.wayland_cmd = strdup(DATADIR "/wsetup.sh");
      |                              ^~~~~~~
src/config.c:292:37: error: expected ')' before string constant
  292 |  config.wayland_cmd = strdup(DATADIR "/wsetup.sh");
      |                                     ^~~~~~~~~~~~~
      |                                     )
src/config.c:296:37: error: expected ')' before string constant
  296 |  config.x_cmd_setup = strdup(DATADIR "/xsetup.sh");
      |                                     ^~~~~~~~~~~~~
      |                                     )
make: *** [makefile:52: obj/src/config.o] Error 1
```

ericonr commented:
You are overwriting the compiler flags, which includes a define for `DATADIR`.  If you want to fix this, commenting out this line https://github.com/nullgemm/ly/blob/master/makefile#L3 from the makefile and adding `export FLAGS="$CFLAGS"` to `pre_build` should be enough.

ericonr commented on srcpkgs/ly/files/ly/run:
```diff
@@ -0,0 +1,7 @@
+
+[ -r conf ] && . ./conf
+
+chvt "${TTY}"
```
I don't love the idea of running `chvt`, but fair enough.

cinerea0 replied:
I tested it without `chvt`, and unfortunately `ly` doesn't switch to its TTY on its own.

ericonr commented on srcpkgs/ly/files/ly/conf:
```diff
@@ -0,0 +1,7 @@
+BAUD_RATE=38400
+TERM_NAME=linux
+TTY=7
```
Perhaps leave a comment here pointing to the service conf?

cinerea0 replied:
Changed!

cinerea0 commented:
Good news: The package now compiles fine on ARM architectures after simplifying the template.
Bad news: The PAM problem still exists.

ericonr commented:
I will try building this locally later to see if I can track anything down.

ericonr commented:
@cinerea0 does this service segfault for you as well? Whenever I log into it I get a segfault in my dmesg.

ericonr commented on srcpkgs/ly/files/ly/run:
```diff
@@ -0,0 +1,7 @@
+
+[ -r conf ] && . ./conf
+
+chvt "${TTY}"
```
This is terrible. If the service is constantly dying for some reason, you can't fix it at all, because it will always run `chvt` again.

cinerea0 replied:
I just tested it again without the `chvt` and it works just fine; I think I might have also changed something else last time. I've removed the `chvt` and `reset` because they're no longer needed.

cinerea0 commented:
@ericonr I didn't think to check my dmesg logs before, but it's segfaulting for me as well. Here's what I believe to be the relevant portion of the logs:
```
[17802.088814] elogind-daemon[1786]: New session 3 of user erebus.
[17802.118293] ly[22191]: segfault at 0 ip 00007f8a457453e1 sp 00007ffc0c1a7248 error 4 in libc-2.30.so[7f8a4560c000+149000]
[17802.118304] Code: 84 00 00 00 00 00 0f 1f 00 31 c0 c5 f8 77 c3 66 2e 0f 1f 84 00 00 00 00 00 89 f9 48 89 fa c5 f9 ef c0 83 e1 3f 83 f9 20 77 1f <c5> fd 74 0f c5 fd d7 c1 85 c0 0f 85 df 00 00 00 48 83 c7 20 83 e1
[17822.337659] elogind-daemon[1786]: Removed session 3.
```
This means that it's not actually a PAM problem but a full-on crashing problem, right?

ericonr commented:
> This means that it's not actually a PAM problem but a full-on crashing problem, right?

It's still potentially a PAM problem, I really don't know. I'm trying to debug it to see if I can catch the error. What seems to happen is that `ly` forks itself to do something, and then that segfaults.

ericonr commented:
Found it! In ly's `env_init` function, it tries to get the contents of the environment variable `LANG`, which might not be set (it isn't, on my system). `getenv` returns a NULL pointer in this case, and when it tries to access that pointer it segfaults. Basically what you need to do is set `LANG` to something, so it doesn't seg fault. I can report this upstream later.

You can do something like this for the service, to solve this:

```
LANG="$LANG" exec setsid ...
```

ericonr commented:
Haven't tested the login service there, but the cross compiled version seems to work fine on armv6l-musl.

ericonr commented:
If you could make a patch / vsed command in `sub/termbox_next/makefile` to use `$(AR)` instead of `ar`, that would be great too. This should also be upstreamed, if possible.

cinerea0 commented:
> If you could make a patch / vsed command in `sub/termbox_next/makefile` to use `$(AR)` instead of `ar`, that would be great too. This should also be upstreamed, if possible.

I added a `vsed` to do this in the template. I also opened a pull request here: https://github.com/nullgemm/termbox_next/pull/5

> LANG="$LANG" exec setsid ...

I added this to the `run` file, and it now works perfectly! Thank you so much! Do you think the `LANG="$LANG"` part could be moved to the `conf` file, or does it have to directly precede the `exec` command?

ericonr commented:
> Do you think the LANG="$LANG" part could be moved to the conf file, or does it have to directly precede the exec command?

It needs to be in the `run` file to be sure it's exported to the application, since the application crashes without it. The user can do whatever they want to it in the `conf` file, and the final value is what ly will see.

cinerea0 commented:
In that case, I think this is fine to to merged, right?

ericonr commented:
My builddir is dirty after the build:

```
argoat-0eb7afae7a001094c9166a8959e021375707522c
configator-8cec1786196ae6f6a8b35e66181277457f2a2bb2
dragonfail-0a2492c6aab3ff64e182db894ce4d40b26799fbd
termbox_next-23fff64470b0730959b39c70aa31fbddd776d9bc
testoasterror-ee7c9d031d4632a6f381a6c174a38539bac04068
```

You can try to use `create_wrksrc` + `build_wrksrc` to make it so this doesn't happen. The `Waybar` template does a similar trick.

ericonr commented on srcpkgs/ly/files/ly/conf:
```diff
@@ -0,0 +1,8 @@
+BAUD_RATE=38400
+TERM_NAME=linux
+# don't change the following value without also changing the "tty" value in /etc/ly/config.ini
+TTY=7
+
+if [ "${TTY}" = "1" ]; then
+	GETTY_ARGS="--noclear"
+fi
```
What do you think about adding an option to source `/etc/locale.conf` ? Having `LANG` set can be nice for your session.

cinerea0 replied:
When you say "option", do you mean putting it in an `if` block like the `GETTY_ARGS`, or having it be commented out by default in the `conf` file with an explanation of what it will do if un-commented?

ericonr replied:
"option" is a bad word. You can just check if it exists, and if so, source it. Kinda like it's done for the `conf` file itself.

cinerea0 replied:
I've added that in the most recent push.

cinerea0 commented:
> My builddir is dirty after the build

I've fixed this in the most recent version of the template. However, while testing`ly`, I ran into a problem that might need fixing (that's why this PR is a draft again). Upon logging in using `ly`, switching to another TTY, and then disabling the `ly` service, the computer freezes. This doesn't happen when using my current DM (`sddm`),so I assume it's unintended behavior. I attempted to fix this by including the `finish` file, but it didn't help. Since the computer freezes, I'm not sure how to check `dmesg` logs from the event, or if those are even the correct logs to check.

ericonr commented:
Does it completely freeze? Is not even `ssh` working?

cinerea0 commented:
> Is not even ssh working?

Do you mean is it possible to ssh into my machine after it freezes? I haven't tried it yet. Might be difficult to do since I'm testing this on my laptop and I don't really have any other devices.

ericonr commented:
There are SSH clients for smartphones, if you have that. Otherwise, can you try listing the steps here? Which session did you log into? I haven't managed to reproduce it.

cinerea0 commented:
I was able to access it through ssh and was able to access `dmesg` logs, but I didn't see anything out of the ordinary. If you can't reproduce it, then it might just be a me problem (and even then, the actions I took aren't likely to occur often enough to be a problem). Assuming that's the case, would it be alright to mark this as ready for review?

ericonr commented:
I think so, it is quite unusual.

Regarding your issue, which session did you start, X or Wayland? When it seems to be locked, could you try pressing `Alt+SysRq+R` and then checking if you can interact with the system? (The SysRq key is usually the Print Screen one - PrtScr)

cinerea0 commented:
I started X. Pressing `Alt+SysRq+R` didn't do anything. Since this problem is likely limited to me, I'll be marking this as ready for review momentarily. If anyone files an issue about it here, asks about it on IRC, or makes a post on the subreddit I'll take another look at it.

Off-topic, but once this gets merged, would you mind taking a look at two other PR's I have open? Travis doesn't report any problems with the most recent version of the templates, so it should be quick: #23924 #23972

cinerea0 commented:
I've edited the opening comment so that merging this PR will automatically close the open issue that's a package request for ly. I would appreciate a maintainer closing the PR I linked in the opening comment after merging this as well. Is there anything remaining that is keeping this from being merged? This (and a couple of my other PRs linked above) have been sitting open for a few days without additional feedback.

PaperMountainStudio commented:
@cinerea0 Hi, I noticed version number is not built in the binary. ly uses git to determine which version it is running, that's why during build there are warnings `/bin/sh: git: command not found`. Adding git into `hostmakedepends` will not solve this, because in the extracted tarball, is not a git repository.

cinerea0 commented:
@PaperMountainStudio In that case, the git errors in that part of the build process can be completely ignored, right?

PaperMountainStudio commented:
You could, but then printing version wouldn't work (not a big issue, I think). Or you could patch one line in the makefile. ---- On Tue, 11 Aug 2020 21:39:04 +0200  notifications@github.com  wrote ----
@PaperMountainStudio In that case, the git errors in that part of the build process can be completely ignored, right?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

cinerea0 commented:
@PaperMountainStudio I've added a `vsed` line to the template in the most recent push that sets the version number correctly.

cinerea0 commented:
After discussion on IRC last night, a verdict was reached that this package is extremely unlikely to be merged in its current state because of all the unversioned dependencies that need to be downloaded and the necessary patching of the configuration file. As such, this PR is currently blocking on https://github.com/nullgemm/ly/issues/225 and https://github.com/nullgemm/ly/issues/226.

abenson commented:
Seemingly abandoned.
